init.lua

```diff
@@ -146,19 +175,30 @@ function cudnn.toDescriptor(t)
    return descriptor
 end
 
+function cudnn.createDescriptors(count, descs_type, create_func, destroy_func)
+   local ds = ffi.new(descs_type, count)
+   for i = 0, count - 1 do
+      errcheck(create_func, ds + i)
+   end
+   local function destroyDescriptors(ds)
+      for i = 0, count - 1 do
+         errcheck(destroy_func, ds[i])
+      end
+   end
+   ffi.gc(ds, destroyDescriptors)
+   return ds
+end
 
 local sharedBuffer = {}
-for i=1,numDevices do
-    sharedBuffer[i] = {}
-end
 
 function cudnn.getSharedWorkspace()
     local device = cutorch.getDevice()
-    local stream = cutorch.getStream() -- starts from 0
-    if not sharedBuffer[device][stream] then
-       sharedBuffer[device][stream] = torch.CudaTensor(1)
+    if not sharedBuffer[device] then
```

Comment:
@borisfom, I don't know if a practise of running convolutions on multiple streams is widespread, but its not hard to imagine situations where it would be useful, so we should not be eliminating support for it . BTW, in some caffe versions dgrad and wgrad are run (potentially) simultaneously on the different streams. 
I don't particularly like allocating almost all free space as a workspace to use FindEx. My preferred approach would be to have a user-specified variable in cudnn to use as maximum workspace size (per gpu per stream, as done now), and a utility function that would calculate the maximum possible workspace that any convolution layer in the net would require (an :apply() that would call cudnnGetWorkspaceSize for all possible algos in a loop). This function may overestimate workspace required, if an algo with the biggest workspace is not the fastest, but hopefully not a big deal - at any rate, not worse than using cudnnFind that tries to allocate exactly that and irreparably fails if it can not. So after propagating a net forward and backward using small workspace and allocating outputs and gradInputs, user can call this function, and set maximum workspace using either its output or some other considerations he can have, that could include potential use of multiple streams. I'm open for discussion on this one, though.